jdaviz/app.py

```diff
@@ -2570,6 +2589,123 @@ def _reset_state(self):
+                for key in layer_dict:
+                    if hasattr(layer_dict[key], "label"):
+                        layer_dict[key] = layer_dict[key].label
+                    elif hasattr(layer_dict[key], "name"):
+                        layer_dict[key] = layer_dict[key].name
+                output_json['viewers'][viewer_id]["layers"][layer.label] = layer_dict
+
+        print(json.dumps(output_json, indent=2))
+
+        with open(output_file, "w") as f:
+            json.dump(output_json, f, indent=2)
+
+    def restore_state(self, input_json=None, save_dir=None):
+        glue_to_plot_option = {
+            "v_min": "stretch_vmin",
+            "v_max": "stretch_vmax",
+            "stretch": "stretch_function",
+            "cmap": "image_colormap",
+            "visible": "image_visible"
+        }
+        if save_dir is None:
+            save_dir = "saved_states"
+        if input_json is None:
+            save_files = glob.glob(f"{save_dir}/glue-state*.json")
+            save_files.sort()
+            input_json=save_files[-1]
+
+        with open(input_json, "r") as f:
+            state_json = json.load(f)
+
+        # Determine if same data was loaded, error if not and we can't find the data on disk.
+        for data_label, origin_dict in state_json["data_origins"].items():
+            if data_label in self.data_collection.labels:
+                continue
+            if origin_dict["file_obj"][:7] == "Object:":
+                print(f"Data loaded from {origin_dict["file_obj"]} is not in data_collection and"
+                      " could not be automatically loaded.")
+            else:
+                # Need to replace null data label with whatever the parser came up with
+                origin_dict["kwargs"]["data_label"] = data_label.split("[")[0]
+                # Load the data!
+                self.load_data(origin_dict["file_obj"],
+                               parser_reference=origin_dict["parser_reference"],
+                               **origin_dict["kwargs"])
+
+        # Create any viewers that do not exist for Imviz
+        if self.config == "imviz":
+            for viewer_id in state_json['viewers']:
+                if viewer_id not in self._viewer_store:
+                    self._jdaviz_helper.create_image_viewer(viewer_id)
+
+        # Load/unload data into appropriate viewers and set plot options
+        plot_options = self._jdaviz_helper.plugins['Plot Options']
+        for viewer_id, viewer in self._viewer_store.items():
+            plot_options.viewer = viewer_id
+            layer_labels = [layer.label for layer in viewer.layers]
+            saved_viewer_state = state_json['viewers'][viewer_id]
+            for layer_id, layer_dict in saved_viewer_state['layers'].items():
+                data_label = layer_id.split(" ")[0]
+                if data_label not in self.data_collection.labels:
+                    print(f"Couldn't find or load {data_label}.")
```

Comment:
I know this is just POC, but hopefully you do not plan to keep this print?

Reply:
I had hoped that would be obvious.

Reply:
The `continue` threw me off. I am not sure if you want some sort of logging here permanently or not.

Reply:
Probably would make this a snackbar message, assuming `restore_state` gets a button in the UI.